### PR TITLE
Remove InternalsVisibleTo from Core to XF.Platforms.*

### PR DESCRIPTION
**Pull request by kingces95**
## Description of Change ###

We hired contractors to build out new backends. They will not live in the XF repo so internal functionality in Core used by our renderers needs to be exposed publicly for use by their renderers. 

Core allows our renderers to access its internal members by applying InternalsVisibleToAttribute (IVT) to itself for each platform assembly. This PR comments out those IVT attributes and fixes the resulting build errors. This is done by either (1) exposing an internal class publicly and changing it's namespace to `Xamarin.Forms.Internals` or (2) exposing a member publicly and adding a `[EditorBrowsable(EditorBrowsableState.Never)]` to the member. This essentially creates a new scope between `internal` and `public` that is expressed as membership in XF.Internals namespace or the presence of `EditorBrowsable` on a public member and is only used by renderer implementors to feed state and events back into core. Customers should expect no support of this scope and should not bind to any of it's members.

The `EditorBrowsable` approach I present as preferable to controller interfaces for hiding members from isense. The attribute is more explicit, doesn't require exposing an additional public interface, nor require changing any of the callsites, is easier to apply, and yet still preserves isesne for the XF team (EditorBrowsable only affects isense of referenced assemblies, not referenced projects). What's not to like! Going forward we can convert the existing controller interfaces to use the attribute approach (recommended) or we could add new controller interfaces on top of the attribute approach as the two are not exclusive. 

IMHO, ultimately, the contractors code should live in the XF repo (maybe in another branch) and each core/platform assembly pair should be merged into a single assembly. This would eliminate this extra scope as well as the need for `ExportRenderer` and forwarder types for the core renderers. This conversion from dynamic to static binding could only speed up our load times. Assembly merging has also been shown to result in smaller assemblies after linking. 

### Bugs Fixed ###

None

### API Changes ###

None. The public surface area we support is unchanged. 

### Behavioral Changes ###

None. The change should result in no behavioral changes. 

### PR Checklist ###

- [X] Has tests (if omitted, state reason in description)
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [X] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/IPinchGestureController.cs [hunk 1]
```diff
@@ -1,6 +1,6 @@
-﻿namespace Xamarin.Forms
+﻿namespace Xamarin.Forms.Internals
 {
-	internal interface IPinchGestureController
+	public interface IPinchGestureController
```
Comment (jassmith) [inline Xamarin.Forms.Core/IPinchGestureController.cs]: Does not need to be in Internals namespace

File: Xamarin.Forms.Core/IPanGestureController.cs [hunk 2]
```diff
@@ -1,6 +1,6 @@
-﻿namespace Xamarin.Forms
+﻿namespace Xamarin.Forms.Internals
 {
-	internal interface IPanGestureController
+	public interface IPanGestureController
```
Comment (jassmith) [inline Xamarin.Forms.Core/IPanGestureController.cs]: Does not need to be in internals namespace

File: Xamarin.Forms.Core/MenuItem.cs [hunk 3]
```diff
@@ -17,7 +18,8 @@ public class MenuItem : BaseMenuItem, IMenuItemController
 
 		public static readonly BindableProperty IconProperty = BindableProperty.Create("Icon", typeof(FileImageSource), typeof(MenuItem), default(FileImageSource));
 
-		internal static readonly BindableProperty IsEnabledProperty = BindableProperty.Create("IsEnabled", typeof(bool), typeof(ToolbarItem), true);
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (jassmith) [inline Xamarin.Forms.Core/MenuItem.cs]: Use controller interface for this class

File: Xamarin.Forms.Core/MultiPage.cs [hunk 4]
```diff
@@ -159,15 +161,17 @@ protected virtual void UnhookContent(T content)
 		{
 		}
 
-		internal static int GetIndex(T page)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static int GetIndex(T page)
```
Comment (jassmith) [inline Xamarin.Forms.Core/MultiPage.cs]: Use controller interface for this class

File: Xamarin.Forms.Core/TapGestureRecognizer.cs [hunk 5]
```diff
@@ -35,7 +36,8 @@ public int NumberOfTapsRequired
 
 		public event EventHandler Tapped;
 
-		internal void SendTapped(View sender)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void SendTapped(View sender)
```
Comment (jassmith) [inline Xamarin.Forms.Core/TapGestureRecognizer.cs]: Use controller interface for this class

File: Xamarin.Forms.Core/VisualElement.cs [hunk 6]
```diff
@@ -86,7 +87,8 @@ public partial class VisualElement : Element, IAnimatable, IVisualElementControl
 
 		public static readonly BindableProperty MinimumHeightRequestProperty = BindableProperty.Create("MinimumHeightRequest", typeof(double), typeof(VisualElement), -1d, propertyChanged: OnRequestChanged);
 
-		internal static readonly BindablePropertyKey IsFocusedPropertyKey = BindableProperty.CreateReadOnly("IsFocused", typeof(bool), typeof(VisualElement), default(bool),
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (jassmith) [inline Xamarin.Forms.Core/VisualElement.cs]: Use controller interface for this class

File: Xamarin.Forms.Core/ListView.cs [hunk 7]
```diff
@@ -210,7 +212,8 @@ public SeparatorVisibility SeparatorVisibility
 			set { SetValue(SeparatorVisibilityProperty, value); }
 		}
 
-		internal ListViewCachingStrategy CachingStrategy { get; private set; }
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (jassmith) [inline Xamarin.Forms.Core/ListView.cs]: Use controller interface for this class

File: Xamarin.Forms.Core/Grid.cs [hunk 8]
```diff
@@ -212,7 +212,8 @@ internal override void ComputeConstraintForView(View view)
 			view.ComputedConstraint = result;
 		}
 
-		internal override void InvalidateMeasureInternal(InvalidationTrigger trigger)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public override void InvalidateMeasureInternal(InvalidationTrigger trigger)
```
Comment (jassmith) [inline Xamarin.Forms.Core/Grid.cs]: Lets make a method on the Controller interface which calls this instead.

File: Xamarin.Forms.Core/Element.cs [hunk 9]
```diff
@@ -158,7 +159,8 @@ internal IPlatform Platform
 		}
 
 		// you're not my real dad
-		internal Element RealParent { get; private set; }
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public Element RealParent { get; private set; }
```
Comment (jassmith) [inline Xamarin.Forms.Core/Element.cs]: Use controller interface for this class

File: Xamarin.Forms.Core/IPanGestureController.cs (hunk 2 above)
Comment (kingces95) [inline Xamarin.Forms.Core/IPanGestureController.cs]: It was moved because it was previously internal. Will move it back to XF namespace and expose publicly. Suggest you review to make sure it's the API you want exposed.

File: Xamarin.Forms.Core/IPinchGestureController.cs (hunk 1 above)
Comment (kingces95) [inline Xamarin.Forms.Core/IPinchGestureController.cs]: It was moved because it was previously internal. Will move it back to XF namespace and expose publicly. Suggest you review to make sure it's the API you want exposed.

File: Xamarin.Forms.Core/MenuItem.cs (hunk 3 above)
Comment (kingces95) [inline Xamarin.Forms.Core/MenuItem.cs]: This is a static member. What convention would you like to use to expose it?

File: Xamarin.Forms.Core/BindableObject.cs [hunk 10]
```diff
@@ -91,7 +91,8 @@ public void SetValue(BindablePropertyKey propertyKey, object value)
 			SetValue(propertyKey.BindableProperty, value, false, false);
 		}
 
-		protected internal static void SetInheritedBindingContext(BindableObject bindable, object value)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static void SetInheritedBindingContext(BindableObject bindable, object value)
 		{
```
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: https://github.com/xamarin/Xamarin.Forms/blob/7cbcd2b2608bb9d2d25d9677417f3e9f9fa44800/Xamarin.Forms.Platform.iOS/Platform.cs#L209 + others

File: Xamarin.Forms.Core/BindableObject.cs [hunk 11]
```diff
@@ -171,7 +172,8 @@ internal bool GetIsDefault(BindableProperty targetProperty)
 			return GetContext(targetProperty) == null;
 		}
 
-		internal object[] GetValues(BindableProperty property0, BindableProperty property1)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public object[] GetValues(BindableProperty property0, BindableProperty property1)
```
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: No references. Revert to internal.

File: Xamarin.Forms.Core/BindableObject.cs [hunk 12]
```diff
@@ -202,7 +204,8 @@ internal object[] GetValues(BindableProperty property0, BindableProperty propert
 			return values;
 		}
 
-		internal object[] GetValues(BindableProperty property0, BindableProperty property1, BindableProperty property2)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public object[] GetValues(BindableProperty property0, BindableProperty property1, BindableProperty property2)
 		{
```
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: https://github.com/xamarin/Xamarin.Forms/blob/7cbcd2b2608bb9d2d25d9677417f3e9f9fa44800/Xamarin.Forms.Platform.iOS/Renderers/FontExtensions.cs#L141 + others

File: Xamarin.Forms.Core/BindableObject.cs [hunk 13]
```diff
@@ -321,12 +324,14 @@ internal void SetValue(BindableProperty property, object value, bool fromStyle)
 			SetValue(property, value, fromStyle, true);
 		}
 
-		internal void SetValueCore(BindablePropertyKey propertyKey, object value, SetValueFlags attributes = SetValueFlags.None)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void SetValueCore(BindablePropertyKey propertyKey, object value, SetValueFlags attributes = SetValueFlags.None)
```
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: No references. Revert to internal.

File: Xamarin.Forms.Core/BindableObject.cs [hunk 14]
```diff
@@ -321,12 +324,14 @@ internal void SetValue(BindableProperty property, object value, bool fromStyle)
...
 		{
 			SetValueCore(propertyKey.BindableProperty, value, attributes, SetValuePrivateFlags.None);
 		}
 
-		internal void SetValueCore(BindableProperty property, object value, SetValueFlags attributes = SetValueFlags.None)
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void SetValueCore(BindableProperty property, object value, SetValueFlags attributes = SetValueFlags.None)
```
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: https://github.com/xamarin/Xamarin.Forms/blob/7cbcd2b2608bb9d2d25d9677417f3e9f9fa44800/Xamarin.Forms.Core/ListView.cs#L277 + many others

File: Xamarin.Forms.Core/MenuItem.cs (hunk 3 above)
Comment (kingces95) [inline Xamarin.Forms.Core/MenuItem.cs]: Actually, there already is a controller. Let me know if the public surface area of that controller is not to your liking.

File: Xamarin.Forms.Core/MultiPage.cs (hunk 4 above)
Comment (kingces95) [inline Xamarin.Forms.Core/MultiPage.cs]: Do you have a preference for how we expose static members? We could just leave them as is for now and if a customer needs those methods just tell them to use the isense hidden ones.

File: Xamarin.Forms.Core/ListView.cs (hunk 7 above)
Comment (kingces95) [inline Xamarin.Forms.Core/ListView.cs]: Already exposed via interface. Really, we should go back and remove the explicit implementations of the controller interfaces from our code, replace them all with the EditorBrowser idiom, and revert all the callsites back to directly accessing the member.

Comment (jassmith) [inline Xamarin.Forms.Core/ListView.cs]: IFooController interfaces are there for when customers need to actually use the methods for implementing renderers. Simply hidden from intellisense is there for when they are not. That is the distinction we need to draw with Controller vs EditorBrowsable

File: Xamarin.Forms.Core/BindableObject.cs (hunk 14 above)
Comment (jassmith) [inline Xamarin.Forms.Core/BindableObject.cs]: This particular call site could just be made to be ((IElementController)Element).SetValueFromRenderer (IsRefreshingProperty, true);

File: Xamarin.Forms.Core/ListView.cs (hunk 7 above)
Comment (kingces95) [inline Xamarin.Forms.Core/ListView.cs]: Yep, got that. I'm suggesting we change how we implement the interfaces (not whether we need them). I'd suggest moving from implicit implementation to explicit implementation with the addition of the EditorBrowser attribute. That way our callsites don't go through the interface and so are cleaner. Customers could do the same but will prob go through the interfaces because that is how they'll get isense as well as know that we're supporting the method.

File: Xamarin.Forms.Core/BindableObject.cs (hunk 14 above)
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: I propose we do a BindableObject scrub in a subsequent PR and keep this a mechanical change. There's a lot of changes here and it'd be good to get them committed as a block with the invariant that no non-trivial control flow changes were made. Then a later investigator could look at the note and dismiss the PR as the source of whatever bug she's looking at.

File: Xamarin.Forms.Core/NumericExtensions.cs [hunk 15]
```diff
@@ -1,14 +1,18 @@
...
+	public static class NumericExtensions
 	{
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
 		public static double Clamp(this double self, double min, double max)
 		{
 			return Math.Min(max, Math.Max(self, min));
 		}
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NumericExtensions.cs]: idem

File: Xamarin.Forms.Core/Performance.cs [hunk 16]
```diff
@@ -1,22 +1,25 @@
...
 using System.Diagnostics;
 using System.Linq;
 using System.Runtime.CompilerServices;
 using System.Text;
 
-namespace Xamarin.Forms
+namespace Xamarin.Forms.Internals
 {
-	internal static class Performance
+	public static class Performance
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Performance.cs]: this doesn't need to be public, it's a tool we use but other implementors might want to use other ones

File: Xamarin.Forms.Core/ReflectionExtensions.cs [hunk 17]
```diff
@@ -1,32 +1,38 @@
...
 using System.Collections.Generic;
+using System.ComponentModel;
 using System.Linq;
 using System.Reflection;
 
-namespace Xamarin.Forms
+namespace Xamarin.Forms.Internals
 {
-	internal static class ReflectionExtensions
+	public static class ReflectionExtensions
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/ReflectionExtensions.cs]: I think this is only used by the Binding system and the Xaml parser. no need to open it. everyone can rewrite the same

File: Xamarin.Forms.Core/VisualElement.cs (hunk 6 above)
Comment (kingces95) [inline Xamarin.Forms.Core/VisualElement.cs]: We agreed statics will just get the Editor attribute.

File: Xamarin.Forms.Core/Element.cs (hunk 9 above)
Comment (kingces95) [inline Xamarin.Forms.Core/Element.cs]: The controller interface declares RealParent so it is on the controller interface.

File: Xamarin.Forms.Core/BindableObject.cs (hunk 12 above)
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: I'd guess the `params` keyword was not used here in order to prevent GC allocation of the array; that signature could be an optimization. For this PR, I'd prefer not to make any control flow or signature modifications to ensure there are no regressions. More to the point, this API was not requested to be put on a controller iface. So this change is not expanding our supported surface area.

File: Xamarin.Forms.Core/Application.cs [hunk 18]
```diff
@@ -48,11 +49,13 @@ public IAppLinks AppLinks
 			}
 		}
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (kingces95) [inline Xamarin.Forms.Core/Application.cs]: Yes

File: Xamarin.Forms.Core/ActionSheetArguments.cs [hunk 19]
```diff
@@ -1,9 +1,9 @@
 using System.Collections.Generic;
 using System.Threading.Tasks;
 
-namespace Xamarin.Forms
+namespace Xamarin.Forms.Internals
 {
-	internal class ActionSheetArguments
+	public class ActionSheetArguments
```
Comment (kingces95) [inline Xamarin.Forms.Core/ActionSheetArguments.cs]: Online doc?

File: Xamarin.Forms.Core/Application.cs [hunk 20]
```diff
@@ -48,11 +49,13 @@ public IAppLinks AppLinks
...
+		public static void SetCurrentApplication(Application value) => Current = value;
+
 		public static Application Current
 		{
 			get { return s_current; }
-			internal set
-			{
+			set {
```
Comment (kingces95) [inline Xamarin.Forms.Core/Application.cs]: Hehe, no such luck...

File: Xamarin.Forms.Core/Color.cs [hunk 21]
```diff
@@ -22,11 +24,14 @@ public static Color Default
 			get { return new Color(-1d, -1d, -1d, -1d, Mode.Default); }
 		}
 
-		internal bool IsDefault
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public bool IsDefault
```
Comment (kingces95) [inline Xamarin.Forms.Core/Color.cs]: I'd rather this PR not change the public surface area exposed to non-platform implementors. Let's leave those for another PR.

File: Xamarin.Forms.Core/Color.cs [hunk 22]
```diff
@@ -22,11 +24,14 @@ public static Color Default
...
 		{
 			get { return _mode == Mode.Default; }
 		}
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static void SetAccent(Color value) => Accent = value;
```
Comment (kingces95) [inline Xamarin.Forms.Core/Color.cs]: Before the EditorBrowsable we would have done that because moving the method to the internals namespace as an extension was the only way to indicate to customers they shouldn't bind to the method. With the browser attribute we can leave the logic where it was originally placed. Not saying it shouldn't be an extension for other reasons but as far as IVT is concerned this change is smaller and gets the job done.

File: Xamarin.Forms.Core/DeviceInfo.cs [hunk 23]
```diff
@@ -2,17 +2,18 @@
 using System.ComponentModel;
 using System.Runtime.CompilerServices;
 
-namespace Xamarin.Forms
+namespace Xamarin.Forms.Internals
 {
-	internal abstract class DeviceInfo : INotifyPropertyChanged, IDisposable
+	public abstract class DeviceInfo : INotifyPropertyChanged, IDisposable
```
Comment (kingces95) [inline Xamarin.Forms.Core/DeviceInfo.cs]: Yeah, good idea. Make 'em even harder to bind to. Let's do that in a separate change PR right after we merge this one.

File: Xamarin.Forms.Core/Grid.cs [hunk 24]
```diff
@@ -212,6 +212,11 @@ internal override void ComputeConstraintForView(View view)
 			view.ComputedConstraint = result;
 		}
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void InvalidateMeasureInernalNonVirtual(InvalidationTrigger trigger)
```
Comment (kingces95) [inline Xamarin.Forms.Core/Grid.cs]: As far as I can tell, the only reason we explicitly implemented the methods on the controller interface was to hide them from isense. Unfortunately, this also had the side effect of breaking all our call sites and requiring us to sprinkle the idiom Foo.Bar -> ((IController)Foo).Bar everywhere. By using the browser attribute we get the desired design effect (hide from isense) but we don't break our call sites which can simply remain Foo.Bar. So I'd actually suggest we revert all explicit implementations and add the browser attribute and at the same time revert the callsites back from ((IController)Foo).Bar to simply Foo.Bar.

File: Xamarin.Forms.Core/NavigationProxy.cs [hunk 25]
```diff
@@ -1,17 +1,19 @@
...
 {
-	internal class NavigationProxy : INavigation
+	public class NavigationProxy : INavigation
 	{
 		INavigation _inner;
 		Lazy<List<Page>> _modalStack = new Lazy<List<Page>>(() => new List<Page>());
 
 		Lazy<List<Page>> _pushStack = new Lazy<List<Page>>(() => new List<Page>());
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (kingces95) [inline Xamarin.Forms.Core/NavigationProxy.cs]: Good point. On the subsequent Internals scrub I'll check to see if adding the attribute to the class simply hides all the members.

File: Xamarin.Forms.Core/NumericExtensions.cs (hunk 15 above)
Comment (kingces95) [inline Xamarin.Forms.Core/NumericExtensions.cs]: Agreed. Could be. But let's save changing the public surface area for another PR. The goal of this PR is just to remove IVT and as we can see that's a pretty big change as is.

File: Xamarin.Forms.Core/Performance.cs (hunk 16 above)
Comment (kingces95) [inline Xamarin.Forms.Core/Performance.cs]: It's in the internals namespace so its not part of our supported public surface.

File: Xamarin.Forms.Core/ReflectionExtensions.cs (hunk 17 above)
Comment (kingces95) [inline Xamarin.Forms.Core/ReflectionExtensions.cs]: Again, in the internals namespace so it's not part of the supported public surface area. It's prob here so it can be shared across all the platforms. If we want to refactor to pull this into each platform or into a utils assembly we can do that in a separate PR.

File: Xamarin.Forms.Core/BindableObject.cs (hunk 14 above)
Comment (kingces95) [inline Xamarin.Forms.Core/BindableObject.cs]: As the API is browser attributed it doesn't constitute part of our supported public surface area and as such could be changed back to internal at a later state. I have no objection to doing SetvalueCore -> SetValueFromRenderer scrub but just not as part of this PR; conversion to SetValueFromRenderer is not necessary for IVT removal and as a non-trivial change could potentially break something.

**Comment by mdizzy** [comment]
Out of curiosity, will this be published in an official release? I'm building some renderers that need things like SendTapped on the TapGestureRecognizer, basically I'm creating irregular shaped views and in order to do so need to implement custom gesture handling.

**Comment by rmarinho** [comment]
@mdizzy is already in public beta 2.3.5-pre
